### DISCUSSION TITLE
Could not log model events for Models without id field

### DISCUSSION THREAD

**Discussion by sishirrijal** [spatie/laravel-activitylog#809] [Q&A]
I am facing a serious issue after including and implementing this package and I am not sure how is this happening.

Environment
---
PHP Version: 7.2.19
Laravel Version: 5.8.29
Package Version: 3.9.1

Scenario
---
I have **table_first** with its model TableFirst with some attributes say: ['id', 't1_first_col', 't1_second_col'].

I have implemented model event logging mechanism in following way.

```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Spatie\Activitylog\Traits\LogsActivity;

class TableFirst extends Model
{
    use LogsActivity;

    protected $table = 'table_first';
    protected static $logFillable = true;
    protected static $logOnlyDirty = true;
    protected static $logName = 'table_first_log';
    protected $fillable = ['t1_first_col', 't1_second_col'];
}
```

I also have **table_second** with a model TableSecond with some attributes: ['table_first_id', 't2_first_col', 't2_second_col']

Model for this table_second is set same as that of first with its corresponding attributes.

i.e.
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Spatie\Activitylog\Traits\LogsActivity;

class TableSecond extends Model
{
    use LogsActivity;

    protected $table = 'table_second';
    protected static $logFillable = true;
    protected static $logOnlyDirty = true;
    protected static $logName = 'table_first_log';
    protected $fillable = ['table_first_id', 't2_first_col', 't2_second_col'];
}
```

This second table is created for storing some data that might be supplementary for first one so I haven't put an id field in this table.

Expected Behaviour
---
When I try to insert data to `TableFirst` & `TableSecond` using model from my respective controllers it should work fine and should create a log instance in database.

Actual Behaviour
---
It inserts the data into `table_first` table i.e. `TableFirst` model instance is created and also logs the fillable attributes of same.
However, for `TableSecond` neither it creates the instance nor logs the fillable attributes, rather, it throws following sort of error.

```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'id' in 'where clause' (SQL: select * from `table_second` where `id` = 0 limit 1)
```

I am damn sure that, It is not working on the models without id attributes as this is happening in every tables with same scenario on my project

Kindly help me figure this issue.

> Excuse me for not mentioning actual table names and model names from my project.

> Excuse me for typo errors and other silly mistakes as well.

Thanks in advance.

**Comment by Gummibeer** [comment]
Hey,

the requirement isn't an `id` column but a primary key - which is, by the way, a requirement by Laravel Eloquent Models in general.
You can change the primary key and it also doesn't have to be an auto-increment column.
```php
protected $primaryKey = 'table_first_id';
protected $keyType = 'int';
public $incrementing = false;
```

In general I can only recommend to have a dedicated PK (in best case AI) because it makes a lot of things a lot easier.
Try to configure your second model properly and everything should work.

**Reply by sishirrijal** [reply]
Thanks for the response @Gummibeer 

In my scenario I have used **TableSecond** as a model of pivot table where `table_first_id` can only be foreign key which in neither of cases can be assigned as primary key.

Laravel Models does not even recommend **(not a requirement!)** to use id in case of pivot models as there are ready to use methods like attach, detach & sync in those cases.

I have made `table_first_id` as an index column which is not compulsorily a primary key.

Anyway it should have a support to log a model without a primary key if it based on model events.

:)

**Reply by Gummibeer** [reply]
Hey, 
okay - that it's a `Pivot` model wasn't clear in the initial comment.
We have a dedicated section in the docs for pivot models.
https://spatie.be/docs/laravel-activitylog/v3/advanced-usage/logging-model-events#logging-on-pivot-models
It's true that you will need a PK in the pivot table. A bit background: the `activities` table has two columns for the subject `subject_id` and `subject_type`. The `subject_id` is the foreign-key column (without DB side reference as it can be referenced to several tables) and the `subject_type` column contains the subject model class name. Both columns together are used by Laravel to resolve the relationship. If you want to log activities for a pivot table we need to be able to identify the table row - at the end we don't care about the name and type of the PK (with some adjustments you can even use UUIDs) - but we need a PK which is also returned by the `$pivot->getKey()`.

**Reply by sishirrijal** [reply]
@Gummibeer 

Thanks for your humble response.

If that is the solution, I need now to prioritize the Models which requires logging and add primary key on them.
